src/memory/host-tlb.c

```diff
@@ -53,29 +114,60 @@ void hosttlb_flush(vaddr_t vaddr) {
   }
 }
 
-void hosttlb_init() {
-  hosttlb_flush(0);
+
+#ifdef CONFIG_RVH 
+
+paddr_t hostvmtlb_lookup(paddr_t gpaddr, int type) {
+  int id = hostvmtlb_idx(gpaddr);
+  const HostVMTLBEntry *e = (type == MEM_TYPE_IFETCH) ?  &hostvmxtlb[id] : 
+    (type == MEM_TYPE_READ) ? &hostvmrtlb[id] : &hostvmwtlb[id];
+
+  if (e->gppn == hostvmtlb_ppn(gpaddr)) {
+    return e->offset + gpaddr; 
+  }
+  return -1;
 }
 
-static paddr_t va2pa(struct Decode *s, vaddr_t vaddr, int len, int type) {
-  if (type != MEM_TYPE_IFETCH) save_globals(s);
-  // int ret = isa_mmu_check(vaddr, len, type);
-  // if (ret == MMU_DIRECT) return vaddr;
-  paddr_t pg_base = isa_mmu_translate(vaddr, len, type);
-  int ret = pg_base & PAGE_MASK;
-  assert(ret == MEM_RET_OK);
-  return pg_base | (vaddr & PAGE_MASK);
+void hosttvmlb_insert(paddr_t gpaddr, paddr_t paddr, int type) {
+  int id = hostvmtlb_idx(gpaddr);
+  HostVMTLBEntry *e = (type == MEM_TYPE_IFETCH) ?  &hostvmxtlb[id] : 
+    (type == MEM_TYPE_READ) ? &hostvmrtlb[id] : &hostvmwtlb[id];
+
+  // Unlike hosttlb_insert, in order to keep the translation uniform, 
+  // we do not translate the paddr to physical address of NEMU.
+  e->offset = paddr - gpaddr;
+  e->gppn = hostvmtlb_ppn(gpaddr);
+}
+
+void hostvmtlb_flush(paddr_t gpaddr) {
+  if (gpaddr == 0) {
+    memset(hostvmtlb, -1, sizeof(hostvmtlb));
+  } else {
+    paddr_t gppn = hostvmtlb_ppn(gpaddr);
+    int idx = hostvmtlb_idx(gpaddr);
+    if (hostvmrtlb[idx].gppn == gppn) hostvmrtlb[idx].gppn = (sword_t)-1;
+    if (hostvmwtlb[idx].gppn == gppn) hostvmwtlb[idx].gppn = (sword_t)-1;
+    if (hostvmxtlb[idx].gppn == gppn) hostvmxtlb[idx].gppn = (sword_t)-1;
+  }
+}
+
+#endif
+
+
+void hosttlb_init() {
+  hosttlb_flush(0);
+#ifdef CONFIG_RVH
+  hostvmtlb_flush(0);
+#endif
 }
 
 __attribute__((noinline))
 static word_t hosttlb_read_slowpath(struct Decode *s, vaddr_t vaddr, int len, int type) {
   paddr_t paddr = va2pa(s, vaddr, len, type);
   word_t data = paddr_read(paddr, len, type, cpu.mode, vaddr);
   if (likely(in_pmem(paddr))) {
-    HostTLBEntry *e = type == MEM_TYPE_IFETCH ?
-      &hostxtlb[hosttlb_idx(vaddr)] : &hostrtlb[hosttlb_idx(vaddr)];
-    e->offset = guest_to_host(paddr) - vaddr;
-    e->gvpn = hosttlb_vpn(vaddr);
+    hosttlb_insert(vaddr, paddr, type);
+    hosttlb_insert(vaddr, paddr, type);
```

Comment:
insert tlb twice?

Reply:
Maybe the problem about merging, I have fixed it.